### PR TITLE
Feature/gustaf 1.1.0

### PR DESCRIPTION
**Pull request by ktrappe**
Added simple paralization of Stellar input and added otion parameter to define number of threads. Cleaned up some more unnecessary ID comparisons.

### CODE REVIEW

File: apps/gustaf/create_stellarmatches_from_file.h [hunk 1]
```diff
@@ -73,6 +73,18 @@ void _getShortIds(StringSet<TId> & shortIds, StringSet<TId> & longIds)
...
+    TSetIterator it = begin(longIds);
+    for(; it != end(longIds); ++it)
+    {
+        TId sId;
+        TIdIterator itId = begin(*it);
+        readUntil(sId, *itId, IsWhitespace());
+        appendValue(shortIds, sId);
+    }
+    */
 }
```
Comment (rrahn) [inline apps/gustaf/create_stellarmatches_from_file.h]: Can this be removed? Maybe document what it means if you might need it later.

File: apps/gustaf/create_stellarmatches_from_file.h [hunk 2]
```diff
@@ -81,157 +93,170 @@ void _getShortIds(StringSet<TId> & shortIds, StringSet<TId> & longIds)
...
 template <typename TSequence, typename TId>
 bool _createStellarMatches(StringSet<TSequence> & queries,
                            StringSet<TId> const & sQueryIds,
                            StringSet<TSequence> & databases,
-                           StringSet<TId> const & sDBIds,
                            StringSet<TId> const & databaseIds,
                            LocalMatchStore<> & lmStore,
-                           StringSet<QueryMatches<StellarMatch<TSequence, TId> > > & stQueryMatches)
+                           StringSet<QueryMatches<StellarMatch<TSequence, TId> > > & stQueryMatches,
+                           unsigned & numThreads)
```
Comment (rrahn) [inline apps/gustaf/create_stellarmatches_from_file.h]: I'll guess numThreads is rdonly? I suggest copy-by-value if you don't change this value within the function.

File: apps/gustaf/create_stellarmatches_from_file.h [hunk 3]
```diff
@@ -81,157 +93,170 @@ void _getShortIds(StringSet<TId> & shortIds, StringSet<TId> & longIds)
...
 {
     typedef typename Infix<TSequence>::Type TInfix;
     typedef typename StellarMatch<TSequence, TId>::TAlign TAlign;
     typedef typename Row<TAlign>::Type TRow;
     typedef typename LocalMatchStore<>::TPosition TPosition;
+    typedef typename LocalMatchStore<>::TMatchStore TMatchStore;
+    typedef typename Iterator<TMatchStore, Standard>::Type TMatchStoreIterator;
+
+    //omp_set_num_threads(4);
```
Comment (rrahn) [inline apps/gustaf/create_stellarmatches_from_file.h]: Can this be removed?

File: apps/gustaf/create_stellarmatches_from_file.h [hunk 4]
```diff
@@ -81,157 +93,170 @@ void _getShortIds(StringSet<TId> & shortIds, StringSet<TId> & longIds)
...
+		bool orientation = true;
+		if ((*it).subjectBeginPos > (*it).subjectEndPos)
+		{
+		    orientation = false;
+		    TPosition tmp = (*it).subjectBeginPos;
+		    (*it).subjectBeginPos = (*it).subjectEndPos;
+		    (*it).subjectEndPos = tmp;
+		}
+		// Computing infices for alignment rows for stellar matches
+		queryInf = infix(queries[iQuery],
```
Comment (rrahn) [inline apps/gustaf/create_stellarmatches_from_file.h]: This could be put in one line.

File: apps/gustaf/create_stellarmatches_from_file.h [hunk 5]
```diff
@@ -81,157 +93,170 @@ void _getShortIds(StringSet<TId> & shortIds, StringSet<TId> & longIds)
...
+				 (*it).queryBeginPos,
+				 (*it).queryEndPos);
+		dbInf = infix(databases[iDB],
```
Comment (rrahn) [inline apps/gustaf/create_stellarmatches_from_file.h]: Same here.

File: apps/gustaf/create_stellarmatches_from_file.h [hunk 6]
```diff
@@ -81,157 +93,170 @@ void _getShortIds(StringSet<TId> & shortIds, StringSet<TId> & longIds)
...
-                         lmStore.matchStore[i].queryEndPos);
-        dbInf = infix(databases[iDB],
-                      lmStore.matchStore[i].subjectBeginPos,
-                      lmStore.matchStore[i].subjectEndPos);
+		// Checking orientation and swapping positions for reverse matches to apply them to stellar format
+		bool orientation = true;
+		if ((*it).subjectBeginPos > (*it).subjectEndPos)
+		{
+		    orientation = false;
+		    TPosition tmp = (*it).subjectBeginPos;
```
Comment (esiragusa) [inline apps/gustaf/create_stellarmatches_from_file.h]: There's a std::swap() function for this...

File: apps/gustaf/create_stellarmatches_from_file.h [hunk 7]
```diff
@@ -81,157 +93,170 @@ void _getShortIds(StringSet<TId> & shortIds, StringSet<TId> & longIds)
...
+			else if (cigar[j].operation == 'D')
+			{
+			    for (unsigned gap = 0; gap < cigar[j].count; ++gap)
+			    {
+				insertGap(row2, gapIndex);
+				++gapIndex;
+			    }
+			}
+			else
+			    gapIndex += cigar[j].count;
```
Comment (rrahn) [inline apps/gustaf/create_stellarmatches_from_file.h]: For easier reading you could also add curly brackets to the last else. 
In general you can leave the brackets if every statement of the if-else if-else block consists of only one line. 
As soon as there is one block with multiple lines you should add the curly brackets to all blocks.

File: apps/gustaf/create_stellarmatches_from_file.h [hunk 8]
```diff
@@ -243,17 +268,13 @@ bool _createStellarMatches(StringSet<TSequence> & queries,
...
 template <typename TSequence, typename TId, typename TMatches>
 bool _getStellarMatchesFromFile(StringSet<TSequence> & queries,
-                                StringSet<TId> & queryIDs,
+                                StringSet<TId> & sQueryIds,
                                 StringSet<TSequence> & databases,
                                 StringSet<TId> & databaseIDs,
                                 CharString const & smFileName,
-                                TMatches & stQueryMatches)
+                                TMatches & stQueryMatches,
+                                unsigned & numThreads)
```
Comment (rrahn) [inline apps/gustaf/create_stellarmatches_from_file.h]: Same here with the reference. Better call-by-value if you don't change it.

File: apps/gustaf/gustaf_matepairs.h [hunk 9]
```diff
@@ -99,15 +99,28 @@ _importSequences(CharString const & fileNameL,
         append(qual, qualR);
         appendValue(seqs, seq, Generous());
         appendValue(quals, qual, Generous());
-        appendValue(ids, id, Generous());
+       // appendValue(ids, id, Generous());
```
Comment (rrahn) [inline apps/gustaf/gustaf_matepairs.h]: Can this be removed?

File: apps/gustaf/gustaf_matepairs.h [hunk 10]
```diff
@@ -99,15 +99,28 @@ _importSequences(CharString const & fileNameL,
...
         clear(qual);
     }
+    
+    // Check for dupliacte id entries.
+    StringSet<TId> uniqueIds = ids;
+    std::sort(begin(uniqueIds, Standard()), end(uniqueIds, Standard()), IdComparator<TId>());  // O(n*log(n))
+    TIdSetIterator itOldEnd = end(uniqueIds, Standard());
+    TIdSetIterator itNewEnd = std::unique(begin(uniqueIds, Standard()), end(uniqueIds, Standard()), IdComparator<TId>());  // O(n)
+
+    --itNewEnd;
```
Comment (rrahn) [inline apps/gustaf/gustaf_matepairs.h]: Why do you decrease it?
I thought itNewEnd points behind the last element not removed which is the same as itOldEnd if nothing was removed.

File: apps/gustaf/gustaf_matepairs.h [hunk 11]
```diff
@@ -99,15 +99,28 @@ _importSequences(CharString const & fileNameL,
...
+    unsigned diff = itOldEnd - itNewEnd;
+    if (length(ids) - diff > 0)
```
Comment (rrahn) [inline apps/gustaf/gustaf_matepairs.h]: Then simply check if itNewEnd != itOldEnd.

File: apps/gustaf/gustaf_matepairs.h [hunk 12]
```diff
@@ -167,10 +179,10 @@ _importSequences(CharString const & fileNameL,
         append(seq, seqR);
         // Note: saving ID of right(!) mate per default
         appendValue(seqs, seq, Generous());
-        appendValue(ids, idR, Generous());
+        //appendValue(ids, idR, Generous());
```
Comment (rrahn) [inline apps/gustaf/gustaf_matepairs.h]: Can this be removed?

File: apps/gustaf/gustaf_matepairs.h [hunk 13]
```diff
@@ -181,15 +193,20 @@ _importSequences(CharString const & fileNameL,
...
     std::sort(begin(uniqueIds, Standard()), end(uniqueIds, Standard()), IdComparator<TId>());  // O(n*log(n))
     TIdSetIterator itOldEnd = end(uniqueIds, Standard());
     TIdSetIterator itNewEnd = std::unique(begin(uniqueIds, Standard()), end(uniqueIds, Standard()), IdComparator<TId>());  // O(n)
 
-    if (itOldEnd - itNewEnd > 0)
-        std::cout << "Found " << itOldEnd - itNewEnd << " nonunique sequence IDs" << std::endl;
-
+    --itNewEnd;
+    unsigned diff = itOldEnd - itNewEnd;
+    if (length(ids) - diff > 0)
```
Comment (rrahn) [inline apps/gustaf/gustaf_matepairs.h]: Same as above.

File: apps/gustaf/msplazer.h [hunk 14]
```diff
@@ -113,7 +113,8 @@ struct MSplazerOptions
         libSize(0),
         libError(0),
         pairedEndMode(false),
-        revCompl(true){}
+        revCompl(true),
+        numThreads(1){}
```
Comment (rrahn) [inline apps/gustaf/msplazer.h]: {} are usually put on a new line ;)

File: apps/gustaf/msplazer_out.h [hunk 15]
```diff
@@ -374,7 +383,8 @@ bool _writeGlobalBreakpoints(String<TBreakpoint> & globalBreakpoints,
             // Write record
             try
             {
-                writeRecord(gffOut, gff_record);
+                //if (tempBP.svtype == TBreakpoint::DISPDUPLICATION)
```
Comment (rrahn) [inline apps/gustaf/msplazer_out.h]: Can this be removed?

File: apps/gustaf/stellar_routines.h [hunk 16]
```diff
@@ -613,7 +614,8 @@ inline bool
 _importSequences(CharString const & fileName,
                  CharString const & name,
                  StringSet<TSequence> & seqs,
-                 StringSet<TId> & ids)
+                 StringSet<TId> & ids
+                 )
```
Comment (rrahn) [inline apps/gustaf/stellar_routines.h]: ) Should not be on a new line.

File: apps/gustaf/stellar_routines.h [hunk 17]
```diff
@@ -636,18 +637,22 @@ _importSequences(CharString const & fileName,
...
-    if (itOldEnd - itNewEnd > 0)
-        std::cout << "Found " << itOldEnd - itNewEnd << " nonunique sequence IDs" << std::endl;
+    StringSet<TId> uniqueIds = ids;
+    std::sort(begin(uniqueIds, Standard()), end(uniqueIds, Standard()), IdComparator<TId>());  // O(n*log(n))
+    TIdSetIterator itOldEnd = end(uniqueIds, Standard());
+    TIdSetIterator itNewEnd = std::unique(begin(uniqueIds, Standard()), end(uniqueIds, Standard()), IdComparator<TId>());  // O(n)
+
+    --itNewEnd;
+    unsigned diff = itOldEnd - itNewEnd;
+    if (length(ids) - diff > 0)
```
Comment (rrahn) [inline apps/gustaf/stellar_routines.h]: Same as some comments before.

**Comment by rrahn** [comment]
Looks good to me. Just a few minor remarks.

File: apps/gustaf/create_stellarmatches_from_file.h (hunk 2 above)
Comment (ktrappe) [inline apps/gustaf/create_stellarmatches_from_file.h]: Changed it. What's the advantage here?

File: apps/gustaf/create_stellarmatches_from_file.h (hunk 7 above)
Comment (ktrappe) [inline apps/gustaf/create_stellarmatches_from_file.h]: Ok

File: apps/gustaf/gustaf_matepairs.h (hunk 10 above)
Comment (ktrappe) [inline apps/gustaf/gustaf_matepairs.h]: I don't know what it does, but it does not work the way we thought. In your version, itNewEnd != itOldEnd was always the case even if there are no nonunique IDs. Also, I couldn't get the number of nonunique IDs right, it was always wrong if there was more than one nonunique ID (especially for different nonunique IDs). I had enough after more than 2 hours, you are welcome to play around with this.

File: apps/gustaf/gustaf_matepairs.h (hunk 11 above)
Comment (ktrappe) [inline apps/gustaf/gustaf_matepairs.h]: see above

**Comment by ktrappe** [comment]
Resolved all comments except for the std::unique issues.

File: apps/gustaf/stellar_routines.h [hunk 18]
```diff
@@ -282,11 +283,11 @@ void _getMatchDistanceScore(
     typedef typename Iterator<TQueryMatchSet, Standard>::Type TQueryMatchSetIterator;
 
     // TODO(rmaerker): Add threads to options and then remove line below or adapt to parallelization scheme.
-    omp_set_num_threads(1);
+    omp_set_num_threads(numThreads);
     Splitter<TQueryMatchSetIterator> setSplitter(begin(queryMatchesSet, Standard()), end(queryMatchesSet, Standard()));
 
     SEQAN_OMP_PRAGMA(parallel for)
-    for(unsigned jobId = 0; jobId < length(setSplitter); ++jobId)
```
Comment (esiragusa) [inline apps/gustaf/stellar_routines.h]: This breaks MSVC builds (see http://cdash.seqan.de/viewBuildError.php?buildid=118518). Either use a signed integer or disable gustaf parallelization on MSVC/OpenMP < 3.0.

Comment (ktrappe) [inline apps/gustaf/stellar_routines.h]: Are there any issues if I change this to int?

**Comment by esiragusa** [comment]
Damn this PR added some new warnings (see http://cdash.seqan.de/viewBuildError.php?type=1&buildid=118585). Could you please fix them asap? We are about to release 2.0.0.

**Comment by rrahn** [comment]
You have to cast the length result from the splitter to an unsigned int as well.

**Comment by rrahn** [comment]
BTW, you don't need to fix it. It will be fixed with #804.

**Comment by esiragusa** [comment]
Ah cool, then let's close this one :+1:

**Pull request merged**